designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,808 @@
+};
+```
+
+ESLint could then be run with this command:
+
+```bash
+eslint src/
+```
+
+When evaluating the `files` array in the config, ESLint will end up searching for `src/**/*.js` and `src/**/*.jsx`. (More information about file resolution is included later this proposal.)
+
+### Replacing `--rulesdir`
+
+In order to recreate the functionality of `--rulesdir`, a user would need to create a new entry in `defs.ruleNamespaces` and then specify the rules from a directory. This can be accomplished using the [`requireindex`](https://npmjs.com/package/requireindex) npm package:
+
+```js
+const requireIndex = require("requireindex");
+
+exports.config = {
+    defs: {
+        ruleNamespaces: {
+            custom: requireIndex("./custom-rules")
+        },
+    }
+    rules: {
+        "custom/my-rule": "error"
+    }
+};
+```
+
+The `requireIndex()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).
+
+### Function Configs
+
+Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:
+
+```js
+exports.config = (context) => {
+
+    // do something
+
+    return {
+        files: ["*.js"],
+        rules: {
+            "semi": ["error", "always"]
+        }
+    };
+
+};
+```
+
+The `context` object has the following members:
+
+* `core` - information about the ESLint core that is using the config
+    * `version` - the version of ESLint being used
+    * `hasRule(ruleId)` - determine if the given rule is in the core
+* `cwd` - the current working directory for ESLint (might be different than `process.cwd()` but always matches `CLIEngine.options.cwd`, see https://github.com/eslint/eslint/issues/11218)
+
+This information allows users to make logical decisions about how the config should be constructed.
+
+A configuration function may return an object or an array of objects. An error is thrown if any other type of value is returned.
+
+#### Checking for Rule Existence
+
+One of the problems with shareable configs today is when a new rule is added to the ESLint core, shareable configs using that rule are not valid for older versions of ESLint (because ESLint validates that configured rules are present). With advanced configs, a shareable config could detect if a new rule is present before deciding to include it, for example:
+
+```js
+const requireIndex = require("requireindex");
+
+exports.config = (context) => {
+    const myConfig = {
+        files: ["*.js"],
+        defs: {
+            ruleNamespaces: {
+                custom: requireIndex("./custom-rules")
+            },
+        }
+        rules: {
+            "custom/my-rule": "error"
+        }
+    };
+
+    if (context.hasRule("some-new-rule")) {
```

Comment:
Could this be simplified by filtering out falsy rules?  e.g.:
```
    rules: {
        "some-new-rule": context.hasRule("some-new-rule") ? ["error"] : null
    }
```
Or, perhaps a helper function:  `context.ruleIfExists('some-new-rule, ["error"])`

Reply:
I feel like we won't really know what helper functions are useful until we get a base implementation out. It'll be easy to see what's missing once we get to see the feedback, whereas we can never remove anything if we add something that isn't needed.